src/main/java/org/truffleruby/core/regexp/EncodingCache.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core.regexp;
+
+import java.util.HashMap;
+import java.util.function.Function;
+
+import org.jcodings.Encoding;
+import org.joni.Regex;
+import org.truffleruby.collections.ConcurrentOperations;
+
+public class EncodingCache {
+
+    private final HashMap<Encoding, Regex> encodings = new HashMap<>();
```

Comment:
Could the type be `Map<Encoding, Regex>` ?

Reply:
It could. Given the small number of encodings likely to be cached and the likely stability of that cache I wonder if we'd really do best with something like Java 9's `java.util.ImmutableCollection.Map` variants as the actual cache.

Reply:
This should be a ConcurrentHashMap or have some form of synchronization.
I think @bjfish meant that the declared type should be just `Map`.